### ISSUE TITLE
Gradus.jl + SpectralFitting.jl = <3

### ISSUE THREAD

**Issue by fjebaker** [astro-group-bristol/Gradus.jl#195]
Using this issue to track what needs to get done for seamless integration of Gradus.jl with SpectralFitting.jl so that the models can easily be used and fit to data.

With #191 and #190 things are in good shape. Transfer functions for given parameters are now fast to compute (from ~20 seconds down to ~4 seconds), which means we don't need to be passing pre-computed tables around, but the user can compute their own dense table in about 10 minutes or so. The transfer function interpolation scheme is also ready, so we just need to provide some helper structures and examples to get everything working together.

Approximate things todo:

- #192
- #193

The goal is to get spectral models working in a way that lets the user choose how to put the model together (whether they want different coronal models or a specific disc geometry).

Then useful to have but not pressing

- #194
- #196 
- #198 
- #199

**Comment by fjebaker** [comment]
@phajy let me know if there's anything I'm missing!

**Comment by fjebaker** [comment]
Okay here's a fully working example of an additive model with a fixed analytic emissivity function:

```julia
struct LineProfile{D,T} <: AbstractTableModel{T,Additive}
    table::D
    K::T
    "Spin" 
    a::T
    θ::T 
    rmin::T 
    rout::T 
end

function LineProfile(
    profile,
    table::Gradus.CunninghamTransferTable;
    K = FitParam(1.0),
    a = FitParam(0.998),
    θ = FitParam(45.0),
    rmin = FitParam(1.0),
    rout = FitParam(100.0, upper_limit = 100.0)
)
    setup = integration_setup(profile, table((get_value(a), get_value(θ))))
    LineProfile((; setup = setup, table = table), K, a, θ, rmin, rout)
end

function SpectralFitting.invoke!(output, domain, model::LineProfile)
    grid = model.table.table((model.a, model.θ))
    rmin = if model.rmin < grid.r_grid[1]
        grid.r_grid[1]
    else
        model.rmin
    end
    Gradus.integrate_lineprofile!(output, model.table.setup, grid, domain; rmin = rmin, rmax = model.rout)
    output
end

# boiler plate to generate the transfer function table
function mapper(a, theta)
    @info a, theta
    m = KerrMetric(1.0, a)
    x = SVector(0.0, 10000.0, deg2rad(theta), 0.0)
    d = ThinDisc(0.0, Inf)

    radii = Gradus.Grids._inverse_grid(Gradus.isco(m) + 1e-2, 100.0, 100)
    @time Gradus.transfer_function_grid(m, x, d, radii; verbose = true)
end

as = collect(range(0.0, 0.998, 4))
thetas = collect(range(0.1, 89.0, 4))
big_grid = [mapper(a, t) for a in as, t in thetas]

table = Gradus.CunninghamTransferTable((as, thetas), big_grid)

# construct the model and set some parameters
model = LineProfile(x -> x^-3, table)
set_value!(model.θ, 40.0)
set_value!(model.a, 0.4)
set_value!(model.rout, 50.0)

domain = collect(range(0.1, 1.4, 100))
output = invokemodel(domain, model)

plot(domain[1:end-1], output)
```

**Comment by fjebaker** [comment]
Low inclinations are poor performing. In the above that's because it's interpolating between 0.1 and 9.0, and at 0.1 the transfer functions are really noisy so it interpolates a negative value...

**Comment by fjebaker** [comment]
Yeah fix is simple, just don't go to `0.1` without also going to `1.0` to get a stable knot.
